zcash_primitives/src/transaction/mod.rs

```diff
@@ -125,12 +126,26 @@ impl TransactionData {
         header
     }
 
-    pub fn freeze(self) -> Transaction {
-        Transaction(self)
+    pub fn freeze(self) -> io::Result<Transaction> {
+        Transaction::from_data(self)
     }
 }
 
 impl Transaction {
+    fn from_data(data: TransactionData) -> io::Result<Self> {
+        let mut tx = Transaction(data, TxId([0; 32]));
+        let mut raw = vec![];
+        tx.write(&mut raw)?;
+        (tx.1)
+            .0
```

Comment:
IMO, this would be a good place to have named fields. For example, if the struct were `pub struct Transaction(data: TransactionData, txid: [u8;32])` then this line would become `tx.txid.copy_from_slice(...)` which I think is a little easier to follow.

Reply:
Sure, I can use named fields instead.